### ISSUE TITLE
Design SQL queries to generate Example 1 from mockup ("Who summary")

### ISSUE THREAD

**Issue by ultimike** [communityinclusion/escoach#11]
Mike will work with Alberto to design and confirm SQL query(s) to generate data table for example 1 from the PDF.

![Screen Shot 2021-04-28 at 10 31 13 AM](https://user-images.githubusercontent.com/1727889/116421488-e0d40f80-a80c-11eb-952f-0e5a481a7863.png)

**Comment by ultimike** [comment]
@escoach similar to #8, I'm going through the process of designing the queries and equations for Example 1 (screenshot above).

I'm having a difficult time determining the "Total number of responses" for the "My team" and Me" columns. Should it be the total number of survey responses that the team and/or individual submitted?

```
SELECT count(*) AS TotalAll,
       sum(case when provider = 'no provider' then 1 else 0 end) AS TotalProvider,
       sum(case when email = 'paul.foos@umb.edu' then 1 else 0 end) AS TotalMe
FROM surveycampaign_results;
```

Or, should it be the total number of responses  where `answer483 = 11659` ("Finding, securing jobs")?

Either way, I'm not 100% sure how each of the "Who" category values should be calculated. As the document says, 

`Number of Hrs/Day = (Number of responses to a specific question / Total number of responses) x 8.`

Taking the example of the "Me" calculation for "A person I support" (where the value is 0:10 in the example), using data from the following query:

```
SELECT count(*) AS TotalWhat,
       sum(case when answer481 = 11640 then 1 else 0 end) AS ApersonAll,
       sum(case when answer481 = 11641 then 1 else 0 end) AS FamilyAll,
       sum(case when answer481 = 11643 then 1 else 0 end) AS DisabilityAll,
       sum(case when answer481 = 11642 then 1 else 0 end) AS SomeoneAll,
       sum(case when answer481 = 11752 then 1 else 0 end) AS OtherAll,
       sum(case when answer481 = 11749 then 1 else 0 end) AS NoneAll,
       sum(case when (answer481 IS NOT NULL AND provider = 'no provider') then 1 else 0 end) AS TotalProvider,
       sum(case when (answer481 = 11640 AND provider = 'no provider') then 1 else 0 end) AS ApersonProvider,
       sum(case when (answer481 = 11641 AND provider = 'no provider') then 1 else 0 end) AS FamilyProvider,
       sum(case when (answer481 = 11643 AND provider = 'no provider') then 1 else 0 end) AS DisabilityProvider,
       sum(case when (answer481 = 11642 AND provider = 'no provider') then 1 else 0 end) AS SomeoneProvider,
       sum(case when (answer481 = 11752 AND provider = 'no provider') then 1 else 0 end) AS OtherProvider,
       sum(case when (answer481 = 11749 AND provider = 'no provider') then 1 else 0 end) AS NoneProvider,
       sum(case when (answer481 IS NOT NULL AND email = 'paul.foos@umb.edu') then 1 else 0 end) AS TotalMe,
       sum(case when (answer481 = 11640 AND email = 'paul.foos@umb.edu') then 1 else 0 end) AS ApersonMe,
       sum(case when (answer481 = 11641 AND email = 'paul.foos@umb.edu') then 1 else 0 end) AS FamilyMe,
       sum(case when (answer481 = 11643 AND email = 'paul.foos@umb.edu') then 1 else 0 end) AS DisabilityMe,
       sum(case when (answer481 = 11642 AND email = 'paul.foos@umb.edu') then 1 else 0 end) AS SomeoneMe,
       sum(case when (answer481 = 11752 AND email = 'paul.foos@umb.edu') then 1 else 0 end) AS OtherMe,
       sum(case when (answer481 = 11749 AND email = 'paul.foos@umb.edu') then 1 else 0 end) AS NoneMe
FROM surveycampaign_results
WHERE answer483 = 11659;
```

The calculation should be: 

`Number of Hrs/Day = (ApersonMe / TotalMe) x 8`
`Number of Hrs/Day = (1 / 7) x 8`
`Number of Hrs/Day = 1.14` (decimal hours)
`Number of Hrs/Day = 1:09` (hh:mm)

Is this correct? (It seems high to me.)

Calculating the rest of the "Me" column results in:

| category | hh:mm |
| ---- | ---- |
| A person... | 1:09 |
| Family... | 3:28 |
| Disability... | 2:17 |
| Someone... | 1:09 |
| Other | 0:00 |
| None | 0:00 |
| Not about... | 7:54 |

The math for the "Not about..." value I used is as follows:

`((Total number of all survey responses) - TotalMe)/(Total number of all survey responses) * 8`
`(521-7)/521*8 = 7.89` (decimal hours)
`7:54` (hh:mm)

Where did I go wrong?

Also - I realize I still have to add all the "who else" stuff, I'm just not there yet.

thanks,
-mike

**Comment by escoach** [comment]
@ultimike Good questions. 
A) The denominator is the total number of responses by either a user, a team, or all depending on which of the three clusters we are looking at.

B) It looks like you need to include "...& finding jobs" in the query?  Example 1 focuses on **WHO** was the interaction with **WHILE finding jobs**. Therefore, the query to determine the number of times paul.foos@... was with a job seekers **while finding jobs** should be something like:

sum(case when (answer481 = 11640 **AND 'Finding, securing jobs'= 1** AND email = 'paul.foos@umb.edu') then 1 else 0 end) AS ApersonMe,   (You might need to call this something like ApersonFindingJobMe).

Then divide this number by the total number of responses by 'paul.foos@...' and multiply per 8 to get hrs/day spent interacting with a person I support while finding jobs.

C) The math in C should be revised based on point B above. However, would it be simpler to do the following? 
"Not about..." = 8 hours - [SUM hours (A person + Family + Disability + .....) ]

D) I know you said you have not worked on adding the "Who else" component yet. But if it is of any help, here is how I would envision the final query for the individual user:
 
sum(case when [(answer481 = 11640 **OR WHO ELSE = 11640**] **AND 'Finding, securing jobs'= 1 AND email = 'paul.foos@umb.edu') then 1 else 0 end) AS ApersonMe.
  
I hope that this helps or let me know.

**Comment by ultimike** [comment]
@escoach Thanks for the responses. To confirm:

1. `The denominator is the total number of responses by either a user, a team, or all depending on which of the three clusters we are looking at.` - so the denominator shouldn't change for anything other than when the "data frame" option is changed, correct? In the data set I currently have (from April 28):

- All data: 521
- My team data ("no provider"): 467
- My data ("paul.foos@***.edu): 17

2. Regarding the "Finding jobs" bit, the query above has a `WHERE answer483 = 11659` clause - this limits the data we're looking at to only those who answered "What" with "Finding, securing jobs" I believe. In the query in my previous comment, the value for "ApersonMe" is 1. 

Based on your previous comment, then the calculation to find the amount of time the paul.foos@***.*** user spent on "A person I support" is:

*  1 / 17 * 8 =  0.47 (00:28)

Is this correct?

3.  For the "Not about..." value, when the query return TotalMe = 7

*  (17-7) / 17 * 8 = 4.71 (04:42)

Is this correct?

Once we get this query/calculation confirmed, I'll add the "who else" stuff.

thanks,
-mike

**Comment by escoach** [comment]
@ultimike I confirm your math for items 1 and 2. Thanks.

The math for item 3 is correct. However, I struggle with it as I am thinking about the effect of adding the "WHO ELSE" element to the query. After you add the "WHO ELSE" in the query, then "TotalMe" most likely will be larger than 7 due to 'duplicate' responses when a user interacts with two different people and, therefore, the user provides a response item to both the WHO and WHO ELSE questions. As a result, the "Not About..." data will be distorted and lose meaning. 

I need to think about it, but probably it makes sense to drop the "NotAbout..." figure from the table for now. Unless you have any suggestions?

**Comment by ultimike** [comment]
For now, to keep things simple(ish), I think we should drop the "Not about..." stuff as well.

Updated queries for this example:

```
/* answer481 is "Who", answer590 is "Who else" */
SELECT count(*) AS TotalAll,
       sum(case when provider = 'no provider' then 1 else 0 end) AS TotalProvider, /* Not sure if this is correct or the value in the next query. */
       sum(case when email = 'paul.foos@umb.edu' then 1 else 0 end) AS TotalMe  /* Not sure if this is correct or the value in the next query. */
FROM surveycampaign_results;
SELECT count(*) AS TotalWhat,
       sum(case when ((answer481 = 11640) OR (answer590 = 12118)) then 1 else 0 end) AS ApersonAll,
       sum(case when ((answer481 = 11641) OR (answer590 = 12119)) then 1 else 0 end) AS FamilyAll,
       sum(case when ((answer481 = 11643) OR (answer590 = 12121)) then 1 else 0 end) AS DisabilityAll,
       sum(case when ((answer481 = 11642) OR (answer590 = 12120)) then 1 else 0 end) AS SomeoneAll,
       sum(case when ((answer481 = 11752) OR (answer590 = 12122)) then 1 else 0 end) AS OtherAll,
       sum(case when ((answer481 = 11749) OR (answer590 = 12123)) then 1 else 0 end) AS NoneAll,
       sum(case when (((answer481 IS NOT NULL) OR (answer590 IS NOT NULL)) AND provider = 'no provider') then 1 else 0 end) AS TotalProvider,
       sum(case when (((answer481 = 11640) OR (answer590 = 12118)) AND provider = 'no provider') then 1 else 0 end) AS ApersonProvider,
       sum(case when (((answer481 = 11641) OR (answer590 = 12119)) AND provider = 'no provider') then 1 else 0 end) AS FamilyProvider,
       sum(case when (((answer481 = 11643) OR (answer590 = 12121)) AND provider = 'no provider') then 1 else 0 end) AS DisabilityProvider,
       sum(case when (((answer481 = 11642) OR (answer590 = 12120)) AND provider = 'no provider') then 1 else 0 end) AS SomeoneProvider,
       sum(case when (((answer481 = 11752) OR (answer590 = 12122)) AND provider = 'no provider') then 1 else 0 end) AS OtherProvider,
       sum(case when (((answer481 = 11749) OR (answer590 = 12123)) AND provider = 'no provider') then 1 else 0 end) AS NoneProvider,
       sum(case when (((answer481 IS NOT NULL) OR (answer590 IS NOT NULL)) AND email = 'paul.foos@umb.edu') then 1 else 0 end) AS TotalMe,
       sum(case when (((answer481 = 11640) OR (answer590 = 12118)) AND email = 'paul.foos@umb.edu') then 1 else 0 end) AS ApersonMe,
       sum(case when (((answer481 = 11641) OR (answer590 = 12119)) AND email = 'paul.foos@umb.edu') then 1 else 0 end) AS FamilyMe,
       sum(case when (((answer481 = 11643) OR (answer590 = 12121)) AND email = 'paul.foos@umb.edu') then 1 else 0 end) AS DisabilityMe,
       sum(case when (((answer481 = 11642) OR (answer590 = 12120)) AND email = 'paul.foos@umb.edu') then 1 else 0 end) AS SomeoneMe,
       sum(case when (((answer481 = 11752) OR (answer590 = 12122)) AND email = 'paul.foos@umb.edu') then 1 else 0 end) AS OtherMe,
       sum(case when (((answer481 = 11749) OR (answer590 = 12123)) AND email = 'paul.foos@umb.edu') then 1 else 0 end) AS NoneMe
FROM surveycampaign_results
WHERE answer483 = 11659;
```

-mike

**Comment by ultimike** [comment]
@escoach See the "Gitlab #11" tab of the [Google Sheet I've shared with you](https://docs.google.com/spreadsheets/d/13T2JrHmX-xpxbkB5wk5z9Br4V5hauoWlL4gy0pg7ZXA/edit#gid=659085311) for an example output from this query. Please confirm that all is correct.

thanks,
-mike

**Comment by ultimike** [comment]
From @escoach comment on #8:

> Tab Gitlab#11: It looks like the query yielded much lower figures than expected? For example, "APersonAll" shows 0:12 hrs/day (Cell B4), which is the result of dividing 14 (APersonAll) by 541 (Total). However, the total from the query for All is actually only 61 (Cell A19). This issue applies to all the cells of the table. As a result, the total hours of the three groups range between 1 and 3 whereas it should be 8 hours.

I've updated the "Tab Gitlab #11" in the Google Sheet so that each value is divided by the proper value - please review.

thanks,  
-mike

**Comment by escoach** [comment]
@ultimike I reviewed Gitlab#11. Thanks for revising the data. The math is correct. However, there is something going on in the data, but I am not sure what is causing it. 
Let's look at "All" as an example: I would expect that the total number of responses for "All" is 61 multiplied by 2 minus the number of responses to "None". That is: (61x2)-18=104. However, the query yielded a total of only 90 (SUM B-G, row 18 in Gitlab#11).
Rationale: As Gitlab#8 shows, 61 users selected "Findings" (Cell C15). Therefore, 61 users must have responded the WHO question and potentially all of 61 of them were asked the WHO Else question. So potentially we should see 122 total answers to the WHO/WHO Else combined. However, 18 users responded "None" to WHO (Gitlab#11, Cell G18) and, therefore, these users were not asked WHO Else. Therefore, the actual total expected number of responses to WHO/WHO Else combined should be 122-18=104.
Yet, the query found only 90 responses to the WHO/WHO Else questions combined.
The NULL data should not be the issue here because these are users who responded the "Finding" question, which in the survey comes AFTER the WHO questions. Users cannot respond the "Finding" question unless they have already responded to the WHO questions.
I will reflect more on this, but I wanted to share in case you see what might be going on.

**Comment by ultimike** [comment]
Hmm - that is interesting. 

I tried adding a `AND ((answer481 IS NOT NULL) OR (answer590 IS NOT NULL))` clause to the WHERE part of the query, it didn't change the data - I figured it might change the TotalAll value, but it did not change any of the query results.

Here's some more data I queried:

- Count of records where answer483=11659 (What=Finding, securing jobs) and answer481 (Who) IS NOT NULL AND answer590 IS NULL (Who else): 21
- Count of records where answer483=11659 (What=Finding, securing jobs) and answer481 (Who) IS NOT NULL AND answer590 = 12123 (Who else = None): 11
- Count of records where answer483=11659 (What=Finding, securing jobs) and answer481 (Who) = 11659 (none): 7
- Count of records where answer483=11659 (What=Finding, securing jobs) and answer590 = 12123 (Who else = None): 11 (same as second query above)

I _think_ that if we do the following, things make sense:

`(TotalAll) * 2 - (count when 481 is not null and 590 is null) - (count when 481 is not null and 590 = None)`

`61 * 2 - 21 - 11 = 90` (which is the same as `SUM(B18:G18)` from the GitLab #11 tab.

Thoughts?
-mike

**Comment by escoach** [comment]
@ultimike yes your math makes sense. Problem solved :)

**Comment by escoach** [comment]
@ultimike From what I understand, it looks good. Thanks.

**Comment by ultimike** [comment]
@mgcobblestone Working to validate the queries in the survey_dashboard module, I ran this case on my local as my personal username (so I have no "my data", and my provider is "Institute for Community Inclusion").

Options on the form I had selected are:

- Who: Any interaction
- Where: (nothing selected)
- What: (nothing selected)
- Timeframe: Up-to-date (default value)

Like #8, this query also appears to be missing the `WHERE answer483 IS NOT NULL` bit, leading to an incorrect TotalAll value. 

According to the [logic on the wiki page](https://github.com/communityinclusion/escoach/wiki/Logic-for-determining-which-Dashboard-query-to-use), we should be seeing the "Who summary", but the results I'm seeing are the "What summary". 

I think I need to update the first part of the logic on the wiki page to:

- If ("Timeframe" = up-to-date), then
  - If (("What" = null) AND ("Who" = null) AND (Where = null)), then display "What summary" (similar to #8).
    - Else If (("What" = null) AND ("Who" = "Any interactions") AND (Where = null)), then display "Who summary" (similar to #11).
      - Else If (("What" = null) AND ("Who" = null) AND ("Where" = "Anyplace")), then display "Where summary" (similar to #17).
Else display "Selected activities" (similar to #19).

Does this make sense?

**Comment by ultimike** [comment]
@mgcobblestone Great progress on this one!

I'm now seeing the "Who summary". 
The "All" numbers match what I'm getting with the manual query.
Both the "Me" and "My Team" numbers are very different than [my manual query](https://docs.google.com/spreadsheets/d/13T2JrHmX-xpxbkB5wk5z9Br4V5hauoWlL4gy0pg7ZXA/edit#gid=0) results. 

-mike

**Comment by mgcobblestone** [comment]
HI @ultimike ,
I'm not sure I follow the new logic.  It appears that in the 2 Else If cases that What = Null OR What = [any value] are enclosed in a set of parens that would make them offset each other and have no impact on the results.  Should the What = [any value] be grouped with on of the other clauses on the right?

**Comment by mgcobblestone** [comment]
@ultimike Besides working out the query logic, the Me and My Teams numbers should be better now.

**Comment by ultimike** [comment]
@mgcobblestone I see your point about the summary logic - I think it can be simplified to:

- If ("Timeframe" = up-to-date), then
  - If (("What" = null) AND ("Who" = null) AND (Where = null)), then display "What summary" (similar to #8).
    - Else If (("Who" = "Any interactions") AND (Where = null)), then display "Who summary" (similar to #11).
      - Else If (("Who" = null) AND ("Where" = "Anyplace")), then display "Where summary" (similar to #17).
        - Else display "Selected activities" (similar to #19).

Thoughts?

Also - I pulled in your commits and didn't see much change in the correctness of the "My team" and "Me" values... I attempted to validate the data from the query above (from my [comment on May 20](https://github.com/communityinclusion/escoach/issues/11#issuecomment-845166376)).

-mike

**Comment by ultimike** [comment]
@mgcobblestone and I screenshared a bit and he pushed up a change to fix the "My team" and "Me" values 😄 

We discussed the updated logic above, and at this point in time, it seems correct. This logic is implemented in the `QueryBuilder::buildQuery()` method.

I'll go ahead and update the [logic wiki page](https://github.com/communityinclusion/escoach/wiki/Logic-for-determining-which-Dashboard-query-to-use) with the updated logic above.

-mike
